### PR TITLE
Convolution

### PR DESCRIPTION
**Pull request by EnLAI111**

### CODE REVIEW

File: solvers/DPGD.py [hunk 1]
```diff
@@ -32,10 +35,15 @@ def set_objective(self, A, reg, y, c, delta, data_fit):
...
     def run(self, callback):
         n, p = self.A.shape
-        DA_inv = np.diff(np.linalg.pinv(self.A), axis=0)
-        DA_invDA_invt = DA_inv @ DA_inv.T
-        DA_invy = DA_inv @ self.y
-        AtA_inv = np.linalg.pinv(self.A.T @ self.A)
+        data = np.array([-np.ones(p), np.ones(p)])
+        diags = np.array([0, 1])
+        D = spdiags(data, diags, p-1, p).toarray()
+        DA_inv = np.linalg.pinv(self.A @ np.linalg.pinv(D))
```
Comment (tomMoral) [inline solvers/DPGD.py:41]: Use close form for `pinv(D)` (todo in a follow up PR)

File: datasets/simulated.py [hunk 2]
```diff
@@ -14,11 +15,11 @@ class Dataset(BaseDataset):
...
-        'n_features': [500],
+        'n_features': [250],
         'n_blocks': [10],
         'mu': [0],
         'sigma': [0.1],
-        'type_A': ['identity', 'random'],
+        'type_A': ['identity', 'random', 'conv'],
         'type_x': ['block', 'sin'],
         'random_state': [27]
     }
```
Comment (tomMoral) [inline datasets/simulated.py:25]:
```suggestion
    }

    # This makes sure that for each solver, we have one simulated dataset that
    # will be compatible in the test_solver.
    test_parameters = {
        'type_A': ['random', 'conv'],
        'n_samples, n_features': [(20, 10)]
    }
```

File: solvers/DPGD.py [hunk 3]
```diff
@@ -52,11 +60,12 @@ def run(self, callback):
...
                 v_old[:] = v
                 v[:] = v_acc
-            v = np.clip(v - stepsize * (DA_invDA_invt @ v - DA_invy),
+            v_tmp, _ = cg(AtA, Aty + np.diff(v, append=0, prepend=0))
+            v = np.clip(v + stepsize * np.diff(v_tmp),
                         -self.reg, self.reg)
             if self.use_acceleration:
                 v_acc[:] = v + (t_old - 1.) / t_new * (v - v_old)
-            u = AtA_inv @ (Aty + np.diff(v, append=0, prepend=0))
+            u, _ = cg(AtA, Aty + np.diff(v, append=0, prepend=0))
```
Comment (tomMoral) [inline solvers/DPGD.py:68]: warm start `cg` (in a follow up PR)

**Review by agramfort — APPROVED** [review]
if we confirm that convolution tests are effectively run then let's merge and iterate later.

**Comment by EnLAI111** [comment]
> if we confirm that convolution tests are effectively run then let's merge and iterate later.

Yes, it has passed the convolution tests.

File: datasets/simulated.py [hunk 4]
```diff
@@ -33,12 +41,43 @@ def __init__(self, n_samples=5, n_features=5, n_blocks=1,
...
 
+    def skip(self):
+        if (self.type_A == 'identity' or self.type_A == 'random') \
+           and self.n_samples != self.n_features:
+            return True, \
+                "samples and features number don't match with type of A"
+        elif self.type_A == 'conv' and self.n_samples - self.n_features < 1:
+            return True, \
+                "samples and features number don't match with type of A"
+        return False, None
```
Comment (tomMoral) [inline datasets/simulated.py:52]: This is not used yet. Can you open an issue on the benchopt/benchopt repo?

File: solvers/ADMM.py [hunk 5]
```diff
@@ -33,22 +34,21 @@ def set_objective(self, A, reg, y, c, delta, data_fit):
...
+                                  matvec=lambda x: self.A.T @ (self.A @ x) -
+                                  gamma * np.diff(np.diff(x),
+                                                  append=0, prepend=0))
 
         while callback(u):
             z_old = z
-            u = np.ravel(A_tmp @ (Aty + np.diff(mu, append=0, prepend=0)
-                                  - gamma * np.diff(z, append=0, prepend=0)))
+            u_tmp = Aty + np.diff(mu, append=0, prepend=0) - \
+                gamma * np.diff(z, append=0, prepend=0)
```
Comment (tomMoral) [inline solvers/ADMM.py:50]: Aviod using `\` in the future, use parenthesis if you need multi-line expressions.

File: solvers/ISTA.py [hunk 6]
```diff
@@ -12,7 +12,7 @@ class Solver(BaseSolver):
     stopping_strategy = 'callback'
 
     # any parameter defined here is accessible as a class attribute
-    parameters = {'alpha': [1.9],
+    parameters = {'alpha': [1., 1.9],
                   'use_acceleration': [False, True]}
```
Comment (tomMoral) [inline solvers/ISTA.py:17]: After looking back at the theory, it seems that we should skip if `alpha > 1 and use_acceleration`.

**Pull request merged**